### PR TITLE
Add SPI (MCP23SXX) devices

### PR DESCRIPTION
**Pull request by Red-M**

### CODE REVIEW

**Comment by jposada202020** [comment]
@Red-M if you could run pylint locally first, you could follow this guide. and run a 
https://learn.adafruit.com/improve-your-code-with-pylint
If you are cloning this and creating a virtual env, you could run `pre-comit run -all-files`, the repository is already configured for that and make life easier :)

**Comment by Red-M** [comment]
Thanks, I've run ` pre-commit` and fixed all the issues there.

File: adafruit_mcp230xx/mcp23xxx.py
Comment (lesamouraipourpre) [inline adafruit_mcp230xx/mcp23xxx.py:25]: These can be removed. They are defined in mcp23sxx.py

File: adafruit_mcp230xx/mcp23s17.py
Comment (lesamouraipourpre) [inline adafruit_mcp230xx/mcp23s17.py:224]: If bit 7 (bank) is set to 1 most of the the _MCP23S17_XXX constants will be wrong including the _MCP23S17_IOCON constant so it can't be unset.

You could either mask out that bit to prevent change and print out a warning or raise an exception. I'm not sure which is the best option.

File: adafruit_mcp230xx/mcp23s08.py
Comment (lesamouraipourpre) [inline adafruit_mcp230xx/mcp23s08.py:32]: There are missing property accessors for some these registers when compared with mcp23s17.py

**Review by lesamouraipourpre — COMMENTED** [review]
_Caveat: I don't have the chips to test with._

These look like some great additions to the library. The bank bit in IOCON in mcp23s17.py is my main area of concern as I believe it will break the library if accidentally set (but without hardware I can't verify.)

The documentation needs an update but that can be done as a separate PR. If you want, I can do that for you just let me know.

File: adafruit_mcp230xx/mcp23s08.py
Comment (Red-M) [reply adafruit_mcp230xx/mcp23s08.py:32]: Yes because the 08 variant doesn't have the 2nd bank so it doesn't need b side variables, this is the same as the MCP23008 implementation but with the SPI device bus instead.

File: adafruit_mcp230xx/mcp23s17.py
Comment (Red-M) [reply adafruit_mcp230xx/mcp23s17.py:224]: I don't have the SPI chips just yet but I can test with real hardware in about a week or so when I get my PCBs in.
If I am understanding this correctly, then the MCP23017 implementation also has this bug.

File: adafruit_mcp230xx/mcp23xxx.py
Comment (Red-M) [reply adafruit_mcp230xx/mcp23xxx.py:25]: Will do.

**Comment by Red-M** [comment]
Thanks for the review, I've made changes based on the feedback, I have a concern regarding the IOCON register value in the I2C implementation then.

File: adafruit_mcp230xx/mcp23s17.py
Comment (Red-M) [reply adafruit_mcp230xx/mcp23s17.py:224]: I think if the 2nd bank is disabled, any register values that are changed for the 2nd bank are not applied if the IOCON register bit 7 is set high because the MCP23017 implementation does the same as the MCP23S17 as far as I know.

**Review by jposada202020 — CHANGES_REQUESTED** [review]
Hello, sorry I am late to the discussion, I think we discuss over discord over the weekend regarding this. 
First Thank you for this: and thanks @lesamouraipourpre for reviewing

Some comments regarding the library, and then a more general kind of approach.

## Library
1. Functionality: Ideal, the library should continue working as it is, we can add things but people `ideally` should use the library the same way. we do not know how many projects are out there with this, and probably we do not want to break them
2. For new libraries, you could follow this guide,this provide some tips and tricks on how to create CircuitPython Libraries. There are some additions to it this week so keep updating it.
https://github.com/adafruit/circuitpython/blob/main/docs/design_guide.rst
3. For new libraries and functionalities it is preferable to include an example showing the new features

## Other options and comments.
We could also include this new support in the community bundle, @kattni please advice regarding this

Either way we need to find a way to test, I could have access to S09/S18 not the S17. But I will need to order them. I am willing to test as I think is a fun project. and if @lesamouraipourpre help us out it could be a good addition.

**Comment by Red-M** [comment]
@jposada202020 That is fine but I've not broken any existing code and the original MCP23SXX library owner is wanting to get this in as well per Adafruit_CircuitPython_MCP23Sxx#1.

These patches only add functionality while retaining the same exact structure to ensure backward compatibility with zero changes required from any downstream projects.

**Comment by jposada202020** [comment]
@Red-M No problem, thanks for this, just trying to put the general ideal out there what is optimal. I have not gone in deep through the code, lesamouraipourpre did a good job there already.

**Comment by Red-M** [comment]
Just getting the docs and examples ready to commit as well so the entire thing is all wrapped up in a neat little package.

**Comment by Red-M** [comment]
Should be another commit coming in soon from my internal version control system.

**Comment by Red-M** [comment]
This should be all good to merge now after reviews and CI testing.

**Comment by Red-M** [comment]
>Either way we need to find a way to test, I could have access to S09/S18 not the S17. But I will need to order them. I am willing to test as I think is a fun project. and if @lesamouraipourpre help us out it could be a good addition.
I actually have s17 chips coming soon and boards to use them on, should be here in about a week.

**Comment by dhalbert** [comment]
> This should be all good to merge now after reviews and CI testing.

> I actually have s17 chips coming soon and boards to use them on, should be here in about a week.

Let's wait to merge until it's tested on actual hardware. I will make this a draft until then so it's not merged accidentally.

**Comment by Red-M** [comment]
>Let's wait to merge until it's tested on actual hardware. I will make this a draft until then so it's not merged accidentally.

No worries, will let you know how well the chips work.

**Comment by Red-M** [comment]
Annoyingly the vendor selling me the SPI variant didn't have their stock listed correctly so its going to be another 3 weeks before I get these chips in.

**Comment by Red-M** [comment]
Anyone mind reviewing this?

**Comment by ladyada** [comment]
@caternuson wanna look after you finish the arduino side?

File: adafruit_mcp230xx/mcp23s17.py
Comment (caternuson) [reply adafruit_mcp230xx/mcp23s17.py:266]: The issue is how the additional registers of the 16 pin variants are addressed. There are two options, which the BANK bit of the IOCON registers determines. This library is hard coded with addresses for BANK=0. So someone could break things by sending in a value that results in the BANK bit being set to 1. I'd suggest checking the input value and raising an exception if that will happen.

File: README.rst
Comment (caternuson) [inline README.rst:17]: Can consolidate this a bit:
```
CircuitPython module for the MCP23017/08 I2C and MCP23S17/08 SPI I/O extenders.
```

**Review by caternuson — CHANGES_REQUESTED** [review]

**Comment by caternuson** [comment]
I'm working on a similar refactor for Arduino. So currently have a setup of each of I2C/SPI and 17/08 = 4 total variants. Tested this library with each for basic output (LED) and input (button). All work fine. Made a few suggestions in code review.

mcp_test

**Comment by Red-M** [comment]
I've used some bitwise operators to just force bit 7 in the io_con register to `0` at all times regardless of user input.

File: adafruit_mcp230xx/mcp23017.py
Comment (caternuson) [inline adafruit_mcp230xx/mcp23017.py:263]: You're wanting to force bit 7 to 0 here? How about:
```python
    @io_control.setter
    def io_control(self, val):
        # force BANK bit to 0
        val &= ~0x80
        self._write_u8(_MCP23017_IOCON, val)
```
But it may be better to raise an exception since someone may think this gives them full access to the IOCON register, and for some reason they want to use BANK=1 mode (although not sure why that'd ever matter). The exception text could say something like "Only BANK=0 is supported".

**Review by caternuson — CHANGES_REQUESTED** [review]

File: adafruit_mcp230xx/mcp23017.py
Comment (Red-M) [reply adafruit_mcp230xx/mcp23017.py:263]: I'm not sure if an exception should even be needed, I don't think its a valid usecase since if you don't need the extra pins, just don't use them.
Just also mentioning that this was also in the original code and is technically a bugfix.

I think adding something to the documentation and forcing the `BANK` bit is enough.

**Comment by caternuson** [comment]
The CI is failing due to an annoying sphinx quirk. It doesn't like the single backticks in the docstring used for code markup. If I remember right, the fix is to change the single backtick to double backticks on both sides. Or can just remove them.

**Comment by caternuson** [comment]
OK. Cool. That fixed. OK, let's go with this as is.

**Pull request merged**
